File: heagen/lib/xronos/fits/xrftgcol.c
/*
 * xrftgcol.c
 */
#include "xronos.h"

int xrftgcol(fitsfile *fptr, XRFile *file, int *status) {
/*
 * Looks at column header keywords to find a match
 * with various strings.  The columns it searches for are:
 *
 *   Quantity               Search string        
 *-----------------------------------------------
 *   Time                   'TIME'               
 *   Rate                   'RATE' or 'COUNT'    
 *   Error                  any substring 'ERR'  
 *   Dead time              'DEADC'              
 *   Integration time       'TIMEDEL'            
 *   Fractional exposure    'FRACEXP'            
 *   PHA                    'PHA'                
 *
 * As an added requirement on the (unitless) fractional exposure the
 * routine checks that the value of the TUNIT keyword for that column is 
 *  blank, = 'none' or does not exist.
 *
 * These defaults can be overridden by the options VX (time)
 * VY (rate), VS (error), VE (dead time), VC (PHA). 
 *
 *  I  fptr  - Pointer to open fits file (current HDU set to data)
 * I/O file  - Contains properties of file
 *  O  status - Error flag (0=OK)
 *
 *  Note: status will always be 0, as found columns get set to their
 *        index and columns that are not found are 0.  No fatal errors
 *        can be encountered in this routine.
 */
   int idum, match, exact;
   char keybuff[FLEN_VALUE];
   char *keyval;
   keyval = keybuff;

   /* Mark CFITSIO errors from this routine for later clearing */
   fits_write_errmark();

   /* Time column*/
   if ( file->VX.set ) {
      file->col_time = file->VX.value;
   } else {
      fits_get_colnum(fptr, CASEINSEN, "TIME", &file->col_time, status);
   }

   /* Rate column (not applicable to event data) */
   if ( file->type == EVENT_DATA ) {
      file->col_rate = 0;
   } else if ( file->VY.set ) {
      file->col_rate = file->VY.value;
   } else {
      *status = 0;
      fits_get_colnum(fptr, CASEINSEN, "RATE*", &file->col_rate, status);
      if ( *status == COL_NOT_FOUND ) {
         *status = 0;
         fits_get_colnum(fptr, CASEINSEN, "COUNT*", &file->col_rate, 
                         status);
      }
   }

   /* Error column (not applicable to event data) */
   if ( file->type == EVENT_DATA ) {
      file->col_err = 0;
   } else if ( file->VS.set ) {
      file->col_err = file->VS.value;
   } else {
      *status = 0;
      fits_get_colnum(fptr, CASEINSEN, "*ERR*", &file->col_err, status);
   }

   /* Dead time correction column (not applicable to event data) */
   if ( file->type == EVENT_DATA ) {
      file->col_deadc = 0;
   } else if ( file->VE.set ) {
      file->col_deadc = file->VE.value;
   } else {
      *status = 0;
      fits_get_colnum(fptr, CASEINSEN, "DEADC*", &file->col_deadc, 
                      status);
   }

   /* Integration time column (TIMEDEL) */
   *status = 0;
   fits_get_colnum(fptr, CASEINSEN, "TIMEDEL*", &file->col_timedel, 
                   status);

   /* FRACEXP column */
   *status = 0;
   fits_get_colnum(fptr, CASEINSEN, "FRACEXP*", &file->col_fracexp, 
                   status);
   /* If FRACEXP exists, ignore DEADC */
   if ( file->col_fracexp > 0 ) file->col_deadc = 0;

   if ( file->col_fracexp ) {
      *status = 0;
      /* If fracexp column has blank or 'NONE' for TUNIT, ignore */
      keyval[0] = '\0';
      fits_read_keys_str(fptr, "TUNIT", file->col_fracexp, 1, &keyval,
                         &idum, status);
      fits_compare_str("NONE", keyval, CASEINSEN, &match, &exact);
      if ( *keyval && !match ) file->col_fracexp = 0;
   }
   
   /* PHA column */
   if ( file->VC.set ) {
      file->col_pha = file->VC.value;
   } else {
      *status = 0;
      fits_get_colnum(fptr, CASEINSEN, "PHA*", &file->col_pha, status);
      if ( *status == 0 ) { /* PHAS not valid */
         fits_get_colnum(fptr, CASEINSEN, "PHAS", &idum, status);
         if ( *status == 0 ) { file->col_pha = 0; }
      }
   }
   *status = 0;
 
   /* Clear CFITSIO errors from this routine */
   fits_clear_errmark();

   return(*status);
}

File: ftools/timing/xronos/xronlib/fits/xrftgcol.f
c
      subroutine xrftgcol(nfield,ttype,tunit,ivect)
      implicit none
c
c Get XRonos COLumns to be read, based on FiTs TTYPEnnn keywords.

c This routine searches the column header keywords to find a match
c with various strings.  The columns it searches for are:

c   Quantity               Search string             Ivect index
c----------------------------------------------------------------
c   Time                   'TIME'                         1
c   Rate                   'RATE' or 'COUNT'              2
c   Error                  any substring 'ERR'            3
c   Dead time              'DEADC'                        4
c   Integration time       'TIMEDEL'                      5
c   Fractional exposure    'FRACEXP'                      6
c   GTI start              any substring 'START'          7
c   GTI stop               any substring 'STOP'           8
c   PHA                    'PHA'                          9

c As an added requirement on the (unitless) fractional exposure the routine 
c checks that the value of the TUNIT keyword for that column is blank, = 'none'
c or does not exist.

c In xronos these defaults can be overridden by the options VX (time)
c VY (rate), VS (error), VE (dead time), VC (PHA).  The special option
c VE0 will cause xronos to ignor the GITS.
c
c   I  nfield   (i) = Total number of columns in the FITS extension.
c   I  ttype(n) (c) = The name of column n.
c   I  tunit(n) (c) = The units of column n.
c   O  ivect    (i) = The column numbers for TIME (1), RATE (2), ERROR (3),
c                      DEADTIME (4), DELTA TIME (5), EXPOSURE (6)
c                    , GTI START (7), GTI STOP (8), PHA (9)

c Author: Eric Lufkin, HEASARC/GSFC, September, 1993
c Revised:                           November, 1993 to add column numbers for
c                                    GTI and EXPOSURE extensions.
c
c NOTE do not initialize ivect in this routine but instead should be done 
c in the calling routine. vector position for colum in gti and rate or event 
c table are in different extension. The ivect(n)=0  within this routine 
c zero the column alredy found. 
c

      character(16) ttype(*),comm,tunit(*)
      integer nfield,ivect(*),n
c      do n=1,9
c         ivect(n) = 0
c      enddo

      DO n = nfield, 1, -1

c Make sure keyword strings are in upper case.

         comm = ttype(n)
         CALL rmvlbk(comm)
         CALL upc(comm)

c Look for various column headers.

         IF   (comm(1:  7).eq.'TIME   ')  ivect(1) = n
         IF  ((comm(1:  4).eq.'RATE'   )  
     &    .or.(comm(1:  5).eq.'COUNT'  )) ivect(2) = n
         IF   (comm(1:  5).eq.'DEADC'  )  ivect(4) = n
         IF   (comm(1:  7).eq.'TIMEDEL')  ivect(5) = n
         IF   (comm(1:  7).eq.'FRACEXP')  ivect(6) = n
         IF  ((comm(1:  3).eq.'PHA'    )
     &    .and.(comm(1: 4).ne.'PHAS'))    ivect(9) = n
c Look for any part of the string that says "ERR".

         IF(index(comm,'ERR').gt.0)   ivect(3) = n

c Look for gti column headers.

         IF(index(comm,'START').gt.0) ivect(7) = n
         IF(index(comm,'STOP').gt.0)  ivect(8) = n

      ENDDO

c Make sure any FRACEXP column is truly a fractional exposure.
c This will be overridden by iopt.

      IF(ivect(6).gt.0) THEN

c >>> Might want a call to xwarn here.<<<

         comm = tunit(ivect(6))
         CALL upc(comm)
         CALL rmvlbk(comm)
         IF((comm.ne.'NONE').and.(comm.ne.' ')) ivect(6) = 0
      ENDIF

      return
      end
